E2E tests failures on Kind 1.12/1.13/1.14

## [1.12.10](https://devops-ci.elastic.co/blue/organizations/jenkins/cloud-on-k8s-e2e-tests-kind-k8s-versions/detail/cloud-on-k8s-e2e-tests-kind-k8s-versions/476/pipeline/42)

Looks like something is wrong with the CRD: `unknown field "subresources" in io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1beta1.CustomResourceDefinitionVersion`

Full log:

```
[2021-07-13T00:47:28.647Z] {"log.level":"info","@timestamp":"2021-07-13T00:47:28.517Z","log.logger":"e2e-ke2en","message":"Creating E2E namespace and role bindings","service.version":"0.0.0-SNAPSHOT+00000000","service.type":"eck","ecs.version":"1.4.0"}
[2021-07-13T00:47:28.909Z] {"log.level":"info","@timestamp":"2021-07-13T00:47:28.711Z","log.logger":"e2e-ke2en","message":"Creating Beat/Agent roles","service.version":"0.0.0-SNAPSHOT+00000000","service.type":"eck","ecs.version":"1.4.0"}
[2021-07-13T00:47:28.910Z] {"log.level":"info","@timestamp":"2021-07-13T00:47:28.904Z","log.logger":"e2e-ke2en","message":"Creating operator namespaces","service.version":"0.0.0-SNAPSHOT+00000000","service.type":"eck","ecs.version":"1.4.0"}
[2021-07-13T00:47:29.171Z] {"log.level":"info","@timestamp":"2021-07-13T00:47:29.080Z","log.logger":"e2e-ke2en","message":"Creating managed namespaces","service.version":"0.0.0-SNAPSHOT+00000000","service.type":"eck","ecs.version":"1.4.0"}
[2021-07-13T00:47:29.433Z] {"log.level":"info","@timestamp":"2021-07-13T00:47:29.254Z","log.logger":"e2e-ke2en","message":"Deploying e2e test secret","service.version":"0.0.0-SNAPSHOT+00000000","service.type":"eck","ecs.version":"1.4.0"}
[2021-07-13T00:47:29.696Z] {"log.level":"info","@timestamp":"2021-07-13T00:47:29.431Z","log.logger":"e2e-ke2en","message":"Installing the Monitoring operator","service.version":"0.0.0-SNAPSHOT+00000000","service.type":"eck","ecs.version":"1.4.0"}
[2021-07-13T00:47:56.314Z] {"log.level":"error","@timestamp":"2021-07-13T00:47:51.755Z","log.logger":"e2e-ke2en","message":"Command execution failed","service.version":"0.0.0-SNAPSHOT+00000000","service.type":"eck","ecs.version":"1.4.0","command":"kubectl apply -f /tmp/eck-e2e/e2e-ke2en/monitoring-operator.yaml","error":"exit status 1"}
[2021-07-13T00:47:56.315Z] namespace/e2e-ke2en configured
[2021-07-13T00:47:56.315Z] serviceaccount/monitoring-elastic-operator created
[2021-07-13T00:47:56.315Z] configmap/monitoring-elastic-operator created
[2021-07-13T00:47:56.315Z] customresourcedefinition.apiextensions.k8s.io/agents.agent.k8s.elastic.co created
[2021-07-13T00:47:56.315Z] error: error validating "/tmp/eck-e2e/e2e-ke2en/monitoring-operator.yaml":
error validating data: [ValidationError(CustomResourceDefinition.spec.versions[0]):
unknown field "subresources" in io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1beta1.CustomResourceDefinitionVersion, ValidationError(CustomResourceDefinition.spec.versions[1]):
unknown field "subresources" in io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1beta1.CustomResourceDefinitionVersion]; if you choose to ignore these errors, turn validation off with --validate=false
[2021-07-13T00:47:56.315Z] 
[2021-07-13T00:47:56.315Z] {"log.level":"info","@timestamp":"2021-07-13T00:47:51.755Z","log.logger":"e2e-ke2en","message":"Deleting resources","service.version":"0.0.0-SNAPSHOT+00000000","service.type":"eck","ecs.version":"1.4.0","file":"/tmp/eck-e2e/e2e-ke2en/secrets.yaml219993989"}
[2021-07-13T00:47:56.315Z] {"log.level":"info","@timestamp":"2021-07-13T00:47:51.842Z","log.logger":"e2e-ke2en","message":"Deleting resources","service.version":"0.0.0-SNAPSHOT+00000000","service.type":"eck","ecs.version":"1.4.0","file":"/tmp/eck-e2e/e2e-ke2en/managed_namespaces.yaml289947966"}
[2021-07-13T00:47:57.705Z] {"log.level":"info","@timestamp":"2021-07-13T00:47:57.385Z","log.logger":"e2e-ke2en","message":"Deleting resources","service.version":"0.0.0-SNAPSHOT+00000000","service.type":"eck","ecs.version":"1.4.0","file":"/tmp/eck-e2e/e2e-ke2en/operator_namespaces.yaml827546779"}
[2021-07-13T00:48:03.009Z] {"log.level":"info","@timestamp":"2021-07-13T00:48:02.653Z","log.logger":"e2e-ke2en","message":"Deleting resources","service.version":"0.0.0-SNAPSHOT+00000000","service.type":"eck","ecs.version":"1.4.0","file":"/tmp/eck-e2e/e2e-ke2en/roles.yaml490473260"}
[2021-07-13T00:48:03.009Z] {"log.level":"info","@timestamp":"2021-07-13T00:48:02.974Z","log.logger":"e2e-ke2en","message":"Deleting resources","service.version":"0.0.0-SNAPSHOT+00000000","service.type":"eck","ecs.version":"1.4.0","file":"/tmp/eck-e2e/e2e-ke2en/rbac.yaml709442241"}
[2021-07-13T00:48:08.310Z] {"log.level":"info","@timestamp":"2021-07-13T00:48:08.184Z","log.logger":"e2e-ke2en","message":"Cleaning up the scratch directory","service.version":"0.0.0-SNAPSHOT+00000000","service.type":"eck","ecs.version":"1.4.0","directory":"/tmp/eck-e2e/e2e-ke2en"}
[2021-07-13T00:48:08.310Z] {"log.level":"error","@timestamp":"2021-07-13T00:48:08.184Z","log.logger":"e2e-ke2en","message":"Failed to run e2e tests","service.version":"0.0.0-SNAPSHOT+00000000","service.type":"eck","ecs.version":"1.4.0","error":"failed to apply monitoring operator manifest: command failed: [kubectl apply -f /tmp/eck-e2e/e2e-ke2en/monitoring-operator.yaml]: exit status 1","error.stack_trace":"github.com/spf13/cobra.(*Command).execute\n\t/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:856\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:974\ngithub.com/spf13/cobra.(*Command).Execute\n\t/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:902\nmain.main\n\t/go/src/github.com/elastic/cloud-on-k8s/test/e2e/cmd/main.go:31\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:225"}
[2021-07-13T00:48:08.311Z] Error: failed to apply monitoring operator manifest: command failed: [kubectl apply -f /tmp/eck-e2e/e2e-ke2en/monitoring-operator.yaml]: exit status 1
[2021-07-13T00:48:08.311Z] exit status 1
```


## [1.13.12](https://devops-ci.elastic.co/blue/organizations/jenkins/cloud-on-k8s-e2e-tests-kind-k8s-versions/detail/cloud-on-k8s-e2e-tests-kind-k8s-versions/476/pipeline/43) and [1.14.10](https://devops-ci.elastic.co/blue/organizations/jenkins/cloud-on-k8s-e2e-tests-kind-k8s-versions/detail/cloud-on-k8s-e2e-tests-kind-k8s-versions/476/pipeline/44)

`TestNameValidation` is the first to be run:

```
[2021-07-13T01:26:38.955Z]         	Error Trace:	utils.go:86
[2021-07-13T01:26:38.955Z]         	Error:      	Received unexpected error:
[2021-07-13T01:26:38.955Z]         	            	map[common.k8s.elastic.co/type]: kibana != <does not have key>, map[kibana.k8s.elastic.co/name]: es-naming-n7nl-kb-xxxxxxxxxxxxxxxxxx != <does not have key>
[2021-07-13T01:26:38.955Z]         	            	github.com/elastic/cloud-on-k8s/test/e2e/test.CheckSelector
[2021-07-13T01:26:38.955Z]         	            		/go/src/github.com/elastic/cloud-on-k8s/test/e2e/test/checks.go:88
[2021-07-13T01:26:38.955Z]         	            	github.com/elastic/cloud-on-k8s/test/e2e/test/kibana.CheckStatus.func1
[2021-07-13T01:26:38.955Z]         	            		/go/src/github.com/elastic/cloud-on-k8s/test/e2e/test/kibana/checks_k8s.go:112
[2021-07-13T01:26:38.955Z]         	            	github.com/elastic/cloud-on-k8s/test/e2e/test.UntilSuccess.func1.1
[2021-07-13T01:26:38.955Z]         	            		/go/src/github.com/elastic/cloud-on-k8s/test/e2e/test/utils.go:83
[2021-07-13T01:26:38.955Z]         	            	github.com/elastic/cloud-on-k8s/pkg/utils/retry.UntilSuccess.func2
[2021-07-13T01:26:38.955Z]         	            		/go/src/github.com/elastic/cloud-on-k8s/pkg/utils/retry/retry.go:41
[2021-07-13T01:26:38.956Z]         	            	runtime.goexit
[2021-07-13T01:26:38.956Z]         	            		/usr/local/go/src/runtime/asm_amd64.s:1371
[2021-07-13T01:26:38.956Z]         	Test:       	TestNameValidation/longestPossibleName/Kibana_status_should_be_updated
[2021-07-13T01:26:38.956Z] {"log.level":"error","@timestamp":"2021-07-13T01:23:45.919Z","message":"stopping early","service.version":"0.0.0-SNAPSHOT+00000000","service.type":"eck","ecs.version":"1.4.0","error":"test failure","error.stack_trace":"github.com/elastic/cloud-on-k8s/test/e2e/test.StepList.RunSequential\n\t/go/src/github.com/elastic/cloud-on-k8s/test/e2e/test/step.go:44\ngithub.com/elastic/cloud-on-k8s/test/e2e.testLongestPossibleName\n\t/go/src/github.com/elastic/cloud-on-k8s/test/e2e/naming_test.go:83\ntesting.tRunner\n\t/usr/local/go/src/testing/testing.go:1193"}
[2021-07-13T01:26:38.956Z]         --- FAIL: TestNameValidation/longestPossibleName/Kibana_status_should_be_updated (1800.00s)
[2021-07-13T01:26:38.956Z] 
[2021-07-13T01:26:38.956Z] === FAIL: test/e2e TestNameValidation/longestPossibleName (1938.64s)
[2021-07-13T01:26:38.956Z] {"log.level":"info","@timestamp":"2021-07-13T00:51:27.283Z","log.logger":"e2e","message":"Test context initialized","service.version":"0.0.0-SNAPSHOT+00000000","service.type":"eck","ecs.version":"1.4.0","context":{"operator":{"name":"e2e-urz9b-operator","namespace":"e2e-urz9b-elastic-system","operator_replicas":1,"managed_namespaces":["e2e-urz9b-mercury","e2e-urz9b-venus"]},"e2e_image":"docker.elastic.co/eck-ci/eck-e2e-tests:732f19bb-e2e-kind-k8s-versions","e2e_namespace":"e2e-urz9b","e2e_service_account":"e2e-urz9b","elastic_stack_version":"7.13.1","log_verbosity":1,"operator_image":"docker.elastic.co/eck-snapshots/eck-operator:1.7.0-SNAPSHOT-2021-07-13-732f19b","operator_image_repo":"docker.elastic.co/eck-snapshots/eck-operator","operator_image_tag":"1.7.0-SNAPSHOT-2021-07-13-732f19b","test_license":"/var/run/secrets/e2e/test-license.json","test_license_pkey_path":"","test_regex":"^Test","test_run":"e2e-urz9b","monitoring_secrets":"/go/src/github.com/elastic/cloud-on-k8s/.ci/monitoring-secrets.json","test_timeout":1800000000000,"auto_port_forwarding":false,"deploy_chaos_job":false,"local":false,"ignore_webhook_failures":false,"ocp_cluster":false,"ocp3_cluster":false,"pipeline":"e2e/kind-k8s-versions","build_number":"476","provider":"kind","clusterName":"kind-476","kubernetes_version":"1.13.12","test_tags":["kind","ipv4"]}}
[2021-07-13T01:26:38.956Z]     --- FAIL: TestNameValidation/longestPossibleName (1938.64s)
```

Status is actually not updated:

```json
{
    "apiVersion": "kibana.k8s.elastic.co/v1",
    "kind": "Kibana",
    "metadata": {
        "annotations": {
            "association.k8s.elastic.co/es-conf": "{\"authSecretName\":\"es-naming-n7nl-kb-xxxxxxxxxxxxxxxxxx-kibana-user\",\"authSecretKey\":\"e2e-urz9b-mercury-es-naming-n7nl-kb-xxxxxxxxxxxxxxxxxx-kibana-user\",\"caCertProvided\":true,\"caSecretName\":\"es-naming-n7nl-kb-xxxxxxxxxxxxxxxxxx-kb-es-ca\",\"url\":\"https://es-naming-n7nl-xxxxxxxxxxxxxxxxxxxxx-es-http.e2e-urz9b-mercury.svc:9200\",\"version\":\"7.13.1\"}",
            "common.k8s.elastic.co/controller-version": "1.7.0-SNAPSHOT"
        },
        "creationTimestamp": "2021-07-13T00:51:27Z",
        "generation": 1,
        "labels": {
            "test-name": "es-naming-n7nl-kb-xxxxxxxxxxxxxxxxxx"
        },
        "name": "es-naming-n7nl-kb-xxxxxxxxxxxxxxxxxx",
        "namespace": "e2e-urz9b-mercury",
        "resourceVersion": "2442",
        "selfLink": "/apis/kibana.k8s.elastic.co/v1/namespaces/e2e-urz9b-mercury/kibanas/es-naming-n7nl-kb-xxxxxxxxxxxxxxxxxx",
        "uid": "74c4d7e8-e374-11eb-af27-0242ac120005"
    },
    "spec": {
        "count": 1,
        "elasticsearchRef": {
            "name": "es-naming-n7nl-xxxxxxxxxxxxxxxxxxxxx",
            "namespace": "e2e-urz9b-mercury"
        },
        "enterpriseSearchRef": {
            "name": ""
        },
        "http": {
            "service": {
                "metadata": {
                    "creationTimestamp": null
                },
                "spec": {}
            },
            "tls": {
                "certificate": {}
            }
        },
        "monitoring": {
            "logs": {},
            "metrics": {}
        },
        "podTemplate": {
            "metadata": {
                "creationTimestamp": null,
                "labels": {
                    "test-name": "es-naming-n7nl-kb-xxxxxxxxxxxxxxxxxx"
                }
            },
            "spec": {
                "containers": null,
                "securityContext": {
                    "fsGroup": 1000,
                    "runAsNonRoot": true,
                    "runAsUser": 1000
                }
            }
        },
        "version": "7.13.1"
    },
    "status": {
        "count": 0
    }
```

The  following error appears several times in the operator logs:

```json
{"log.level":"error","@timestamp":"2021-07-13T00:51:30.353Z","log.logger":"manager.eck-operator.controller.kb-es-association-controller","message":"Reconciler error","service.version":"1.7.0-SNAPSHOT+732f19bb","service.type":"eck","ecs.version":"1.4.0","name":"es-naming-n7nl-kb-xxxxxxxxxxxxxxxxxx","namespace":"e2e-urz9b-mercury","error":"the server could not find the requested resource (put kibanas.kibana.k8s.elastic.co es-naming-n7nl-kb-xxxxxxxxxxxxxxxxxx)","error.stack_trace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.2/pkg/internal/controller/controller.go:253\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.2/pkg/internal/controller/controller.go:214"}
```

--- a/config/crds/v1beta1/all-crds.yaml
+++ b/config/crds/v1beta1/all-crds.yaml
@@ -347,6 +347,12 @@ spec:
     - apm
     singular: apmserver
   scope: Namespaced
+  subresources:
+    scale:
+      labelSelectorPath: .status.selector
+      specReplicasPath: .spec.count
+      statusReplicasPath: .status.count
+    status: {}
   validation:
     openAPIV3Schema:
       description: ApmServer represents an APM Server resource in a Kubernetes cluster.
@@ -916,17 +922,9 @@ spec:
   - name: v1
     served: true
     storage: true
-    subresources:
-      scale:
-        labelSelectorPath: .status.selector
-        specReplicasPath: .spec.count
-        statusReplicasPath: .status.count
-      status: {}
   - name: v1beta1
     served: true
     storage: false
-    subresources:
-      status: {}
   - name: v1alpha1
     served: false
     storage: false
@@ -3323,6 +3321,12 @@ spec:
     - ent
     singular: enterprisesearch
   scope: Namespaced
+  subresources:
+    scale:
+      labelSelectorPath: .status.selector
+      specReplicasPath: .spec.count
+      statusReplicasPath: .status.count
+    status: {}
   validation:
     openAPIV3Schema:
       description: EnterpriseSearch is a Kubernetes CRD to represent Enterprise Search.
@@ -3836,17 +3840,9 @@ spec:
   - name: v1
     served: true
     storage: true
-    subresources:
-      scale:
-        labelSelectorPath: .status.selector
-        specReplicasPath: .spec.count
-        statusReplicasPath: .status.count
-      status: {}
   - name: v1beta1
     served: true
     storage: false
-    subresources:
-      status: {}
 status:
   acceptedNames:
     kind: ""
@@ -3888,6 +3884,12 @@ spec:
     - kb
     singular: kibana
   scope: Namespaced
+  subresources:
+    scale:
+      labelSelectorPath: .status.selector
+      specReplicasPath: .spec.count
+      statusReplicasPath: .status.count
+    status: {}
   validation:
     openAPIV3Schema:
       description: Kibana represents a Kibana resource in a Kubernetes cluster.
@@ -4532,17 +4534,9 @@ spec:
   - name: v1
     served: true
     storage: true
-    subresources:
-      scale:
-        labelSelectorPath: .status.selector
-        specReplicasPath: .spec.count
-        statusReplicasPath: .status.count
-      status: {}
   - name: v1beta1
     served: true
     storage: false
-    subresources:
-      status: {}
   - name: v1alpha1
     served: false
     storage: false

--- a/config/crds/v1beta1/patches/apm-kibana-patches.yaml
+++ b/config/crds/v1beta1/patches/apm-kibana-patches.yaml
@@ -8,6 +8,12 @@
   path: /spec/validation
 - op: remove
   path: /spec/versions/1/schema
+# move subresources to the top level, see https://github.com/elastic/cloud-on-k8s/issues/4635
+- op: move
+  from: /spec/versions/0/subresources
+  path: /spec/subresources
+- op: remove
+  path: /spec/versions/1/subresources
 # Move additionalPrinterColumns of the v1 version to the top-level, and remove it from the v1beta1 version.
 # This is in order to be compatible with k8s 1.11.
 - op: move

--- a/config/crds/v1beta1/patches/ent-patches.yaml
+++ b/config/crds/v1beta1/patches/ent-patches.yaml
@@ -5,6 +5,13 @@
 - op: remove
   path: /spec/validation/openAPIV3Schema/type
 
+# move subresources to the top level, see https://github.com/elastic/cloud-on-k8s/issues/4635
+- op: move
+  from: /spec/versions/0/subresources
+  path: /spec/subresources
+- op: remove
+  path: /spec/versions/1/subresources
+
 # Using `kubectl apply` stores the complete CRD file as an annotation,
 # which may be too big for the annotations size limit.
 # One way to mitigate this problem is to remove the (huge) podTemplate properties from the CRD.

--- a/deploy/eck-operator/charts/eck-operator-crds/templates/all-crds-legacy.yaml
+++ b/deploy/eck-operator/charts/eck-operator-crds/templates/all-crds-legacy.yaml
@@ -362,6 +362,12 @@ spec:
     - apm
     singular: apmserver
   scope: Namespaced
+  subresources:
+    scale:
+      labelSelectorPath: .status.selector
+      specReplicasPath: .spec.count
+      statusReplicasPath: .status.count
+    status: {}
   validation:
     openAPIV3Schema:
       description: ApmServer represents an APM Server resource in a Kubernetes cluster.
@@ -931,17 +937,9 @@ spec:
   - name: v1
     served: true
     storage: true
-    subresources:
-      scale:
-        labelSelectorPath: .status.selector
-        specReplicasPath: .spec.count
-        statusReplicasPath: .status.count
-      status: {}
   - name: v1beta1
     served: true
     storage: false
-    subresources:
-      status: {}
   - name: v1alpha1
     served: false
     storage: false
@@ -3362,6 +3360,12 @@ spec:
     - ent
     singular: enterprisesearch
   scope: Namespaced
+  subresources:
+    scale:
+      labelSelectorPath: .status.selector
+      specReplicasPath: .spec.count
+      statusReplicasPath: .status.count
+    status: {}
   validation:
     openAPIV3Schema:
       description: EnterpriseSearch is a Kubernetes CRD to represent Enterprise Search.
@@ -3875,17 +3879,9 @@ spec:
   - name: v1
     served: true
     storage: true
-    subresources:
-      scale:
-        labelSelectorPath: .status.selector
-        specReplicasPath: .spec.count
-        statusReplicasPath: .status.count
-      status: {}
   - name: v1beta1
     served: true
     storage: false
-    subresources:
-      status: {}
 status:
   acceptedNames:
     kind: ""
@@ -3933,6 +3929,12 @@ spec:
     - kb
     singular: kibana
   scope: Namespaced
+  subresources:
+    scale:
+      labelSelectorPath: .status.selector
+      specReplicasPath: .spec.count
+      statusReplicasPath: .status.count
+    status: {}
   validation:
     openAPIV3Schema:
       description: Kibana represents a Kibana resource in a Kubernetes cluster.
@@ -4577,17 +4579,9 @@ spec:
   - name: v1
     served: true
     storage: true
-    subresources:
-      scale:
-        labelSelectorPath: .status.selector
-        specReplicasPath: .spec.count
-        statusReplicasPath: .status.count
-      status: {}
   - name: v1beta1
     served: true
     storage: false
-    subresources:
-      status: {}
   - name: v1alpha1
     served: false
     storage: false

